src/main/java/org/zalando/nakadi/util/TimeLogger.java

```diff
@@ -0,0 +1,130 @@
+import java.util.LinkedList;
+
+public class TimeLogger {
+    private static class LogCollector {
+        private long[] stepStarts;
+        private String[] stepNames;
+        private int currentStep;
+
+        LogCollector() {
+            this.stepStarts = new long[10];
+            this.stepNames = new String[this.stepStarts.length];
+            this.currentStep = 0;
+        }
+
+        private void reset() {
+            this.currentStep = 0;
+        }
+
+        void addRecord(final String message) {
+            if (currentStep >= this.stepStarts.length) {
+                this.stepStarts = Arrays.copyOf(this.stepStarts, this.stepStarts.length + 10);
+                this.stepNames = Arrays.copyOf(this.stepNames, this.stepStarts.length);
+            }
+            this.stepStarts[currentStep] = System.currentTimeMillis();
+            this.stepNames[currentStep] = message;
+            currentStep++;
+        }
+
+        String describe() {
+            final StringBuilder builder = new StringBuilder("[" + stepNames[0] + "]");
+            for (int i = 1; i < currentStep - 1; ++i) {
+                builder.append(" [")
+                        .append(stepNames[i])
+                        .append("=")
+                        .append(stepStarts[i + 1] - stepStarts[i])
+                        .append("]");
+            }
+            return builder.toString();
+        }
+    }
+
+    private static final ThreadLocal<LogCollector[]> THREAD_COLLECTORS = new ThreadLocal<>();
+    private static final LinkedList<LogCollector> COLLECTOR_POOL = new LinkedList<>();
+
+    public static void startMeasure(final String heading, final String stepName) {
+        LogCollector collector;
+        synchronized (COLLECTOR_POOL) {
+            collector = COLLECTOR_POOL.poll();
+        }
+        if (null == collector) {
+            collector = new LogCollector();
+        }
+        LogCollector[] collectors = THREAD_COLLECTORS.get();
+        final int idxToSet;
+        if (null == collectors) {
+            collectors = new LogCollector[2];
+            THREAD_COLLECTORS.set(collectors);
+            idxToSet = 0;
+        } else {
+            int freeSlot = -1;
+            for (int idx = 0; idx < collectors.length; ++idx) {
+                if (collectors[idx] == null) {
+                    freeSlot = idx;
+                    break;
+                }
+            }
+            if (freeSlot < 0) {
+                idxToSet = collectors.length;
+                collectors = Arrays.copyOf(collectors, collectors.length + 2);
+                THREAD_COLLECTORS.set(collectors);
+            } else {
+                idxToSet = freeSlot;
+            }
+        }
+        if (idxToSet > 0) {
+            collectors[idxToSet].addRecord(heading);
+        }
+        collectors[idxToSet] = collector;
+        collector.addRecord(heading);
+        collector.addRecord(stepName);
+    }
+
+    public static String finishMeasure() {
```

Comment:
Since the threads are going to be reused by servlet container, maybe it makes sense to clean thread local after we finish all measurements.